### ISSUE TITLE
[Bug] [Portability][MSVC]: M_LOG2E is unavailable when building Flash Attention with NVCC and MSVC

### ISSUE THREAD

**Issue by rogeroberg** [vllm-project/vllm#54459]
### Environment

- Windows Server 2025
- Visual Studio Build Tools with MSVC
- CUDA/NVCC 13.2
- Python 3.12
- PyTorch 2.11.0+cu130
- SystemPanic/vllm-windows 0.25-based deployment
- DeepSeek-V4-Flash-0731
- 4 x H200 NVL GPUs, SM90a

### Background

I encountered this issue while building vllm-project/flash-attention through SystemPanic/vllm-windows for a deployment serving DeepSeek-V4-Flash-0731 on 4 x H200 NVL GPUs.

### Problem

Several Flash Attention CUDA source files use `M_LOG2E` in softmax and `exp2f` scale calculations:

    csrc/flash_attn/flash_api.cpp
    csrc/flash_attn/src/softmax.h
    hopper/flash_bwd_preprocess_kernel.h
    hopper/mainloop_fwd_sm80.hpp
    hopper/mainloop_fwd_sm90_tma_gmma_ws.hpp
    hopper/mainloop_bwd_sm80.hpp
    hopper/mainloop_bwd_sm90_tma_gmma_ws.hpp
    hopper/softmax.h

Examples include:

    params.scale_softmax_log2 =
        softmax_scale * M_LOG2E;

and:

    const float max_scaled =
        max(mi) == -INFINITY
            ? 0.f
            : max(mi) * (
                  Scale_max ? scale : float(M_LOG2E));

`M_LOG2E` represents `log2(e)` and is commonly available in GNU/Linux build environments. Its availability is implementation-specific, however, and MSVC normally exposes the `M_*` constants only when `_USE_MATH_DEFINES` is defined before the relevant math header is included.

I understand that native Windows may not be part of the supported build matrix. This appears to be a small source-portability issue rather than a request for broader Windows support.

### Tested workaround

The working downstream build replaces `M_LOG2E` with the corresponding float constant:

    - float(M_LOG2E)
    + 1.4426950408889634f

For example:

    - params.scale_softmax_log2 =
    -     softmax_scale * M_LOG2E;
    + params.scale_softmax_log2 =
    +     softmax_scale * 1.4426950408889634f;

The replacement preserves the intended `log2(e)` scale used by the float-based softmax and `exp2f` paths.

Applying the replacement to all affected occurrences allowed the Flash Attention extensions to build successfully with NVCC and MSVC.

### Related PyTorch-side solution

The discussion in: https://github.com/vllm-project/vllm/pull/54007

indicates that a PyTorch-side update may resolve this compatibility issue in the Windows build environment and perhaps could the same change be done here as well.

### Other possible solutions

The affected NVCC and MSVC targets could define:

    _USE_MATH_DEFINES=1

For example:

    target_compile_definitions(
        <target> PRIVATE _USE_MATH_DEFINES=1
    )

Alternatively, Flash Attention could use a project-owned `constexpr` value for `log2(e)` instead of depending on the host math-header configuration.

The source-level constant would be independent of PyTorch headers, build flags, and math-header include order.

### Repository context

Seven affected files are inherited from Dao-AILab/flash-attention.

The vllm-project fork additionally contains an affected occurrence in:

    hopper/softmax.h

I am reporting the complete set here because this is the Flash Attention fork built by vLLM.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by Blade-of-Mikasa** [comment]
I'd like to work on this portability fix in `vllm-project/flash-attention`, where the affected sources and FA2/FA3 CMake targets live. I found no open PR there for `M_LOG2E`/MSVC. I plan to add `_USE_MATH_DEFINES` only to the MSVC builds of both extension targets, leaving non-MSVC compile flags unchanged, and reference this issue from the PR.
